### ISSUE TITLE
Allow unsafe virtual-docs

### ISSUE THREAD

**Issue by timothyhinrichs** [open-policy-agent/opa#210]
Sometimes I find myself writing virtual-documents that are technically unsafe but are safe the way I use them.  It would be nice if we could declare a virtual-doc unsafe, and have the evaluation work properly so long as the document is always provided with ground arguments when evaluated.

For example, suppose we are declaring which API paths are permitted.

```
# permit all subpaths of 'foo' that are not explicitly prohibited
permit :- 
    request.path[0] = "foo", 
    not prohibited[request.path]

prohibited[path] :-
    path[0] = "bar",
    path[1] = "baz"
```

Here 'prohibited' is unsafe, but because we will only ever be evaluating 'permit', we see that 'prohibited' will never be problematic for evaluation.  If something ever did ask for 'prohibited' directly, we could return an error message.

We can think of these unsafe virtual-docs as user-written builtins.  Their arguments (or maybe more generally some of their arguments) must be ground before evaluation.

**Comment by tsandall** [comment]
I wonder if this can be handled using the "with" keyword.

For instance, in the example above we would rewrite prohibited as `prohibited :- input.path[0] = "bar", input.path[1] = "baz"`. To apply to arbitrary values, the "with" keyword could be used: `not prohibited with input.path as ["foo", "bar"]`. Note, the value could come from input, data, literals, etc.

One issue with this suggestion is that (currently) we return an error if the caller supplies an input document to query that already defines one using the "with" keyword. This would need to be relaxed.

**Comment by tsandall** [comment]
Fixed by #345
